File: C++/src/ClosestPair2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <cmath>

#define INF ((int) 1e9)

using namespace std;

struct dPoint {
public:
    dPoint(double x, double y) : x(x), y(y) {}

    double x, y;

    double distance(dPoint *b) {
        return sqrt(pow(b->x - this->x, 2) + pow(b->y - this->y, 2));
    }
};

struct Tuple {
    dPoint a, b;

    Tuple(const dPoint &a, const dPoint &b) : a(a), b(b) {}
};

Tuple getClosestPairRec(vector<dPoint> *Px, vector<dPoint> *Py, int start, int stop) {
    if (stop - start <= 3) {
        Tuple *min = nullptr;
        double minvalue = INF;
        for (int i = start; i < stop; i++)
            for (int j = i + 1; j < stop; j++)
                if (Px->at(i).distance(&Px->at(j)) < minvalue) {
                    minvalue = Px->at(i).distance(&Px->at(j));
                    min = new Tuple(Px->at(i), Px->at(j));
                }

        return *min;
    }

    int midPoint = ((stop - start) / 2) + start;
    Tuple minQ = getClosestPairRec(Px, Py, start, midPoint);
    Tuple minR = getClosestPairRec(Px, Py, midPoint, stop);

    auto delta = min(minQ.a.distance(&minQ.b), minR.a.distance(&minR.b));
    auto L = Px->at(midPoint);

    vector<dPoint> S;
    for (int i = start; i < stop; i++) {
        if (L.distance(&Py->at(i)) <= delta) S.push_back(Py->at(i));
    }

    Tuple *minS = nullptr;
    double minvalue = INF;
    for (int i = 0; i < S.size(); i++) {
        for (int j = i + 1; j <= i + 16 && j < S.size(); j++)
            if (S.at(i).distance(&S.at(j)) < minvalue) {
                minvalue = S.at(i).distance(&S.at(j));
                minS = new Tuple(S.at(i), S.at(j));
            }
    }

    if (minS != nullptr && minS->a.distance(&minS->b) < delta)
        return *minS;
    else if (minQ.a.distance(&minQ.b) < minR.a.distance(&minR.b))
        return minQ;
    else return minR;
}

Tuple getClosestPair(vector<dPoint> P) {
    vector<dPoint> Px = P;
    vector<dPoint> Py = P;
    sort(Px.begin(), Px.end(), [](const dPoint &a, const dPoint &b) { return (a.x < b.x); });
    sort(Py.begin(), Py.end(), [](const dPoint &a, const dPoint &b) { return (a.y < b.y); });

    return getClosestPairRec(&Px, &Py, 0, Px.size());
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int testCase;
    while (true) {
        cin >> testCase;
        if (testCase == 0) break;

        vector<dPoint> P;
        double a, b;
        for (int i = 0; i < testCase; i++) {
            cin >> a >> b;
            P.push_back(*new dPoint(a, b));
        }
        auto r = getClosestPair(P);
        cout << r.a.x << " " << r.a.y << " " << r.b.x << " " << r.b.y << endl;
    }
}

File: Java/src/main/java/ClosestPair2.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;

public class ClosestPair2 {
    public static dPoint[] gP;

    public static void main(String[] args) {
        Kattio kattio = new Kattio(System.in, System.out);

        for (int testcase = kattio.getInt(); testcase != 0; testcase = kattio.getInt()) {
            dPoint[] P = new dPoint[testcase];
            for (int i = 0; i < testcase; i++) P[i] = new dPoint(kattio.getDouble(), kattio.getDouble());
            System.out.println(getClosestPair(P));
        }
    }

    public static Tuple<dPoint> getClosestPair(dPoint[] P) {
        gP = P;
        dPoint[] Px = P.clone();
        dPoint[] Py = P.clone();
        Arrays.sort(Px, Comparator.comparingDouble(a -> a.x));
        Arrays.sort(Py, Comparator.comparingDouble(a -> a.y));

        return getClosestPairRec(Px, Py);
    }

    private static Tuple<dPoint> getClosestPairRec(dPoint[] Px, dPoint[] Py) {
        if (Px.length <= 3) {
            Tuple<dPoint> min = null;
            double minvalue = Double.MAX_VALUE;
            for (int i = 0; i < Px.length; i++) {
                for (int j = i + 1; j < Px.length; j++)
                    if (Px[i].distance(Px[j]) < minvalue){
                        minvalue = Px[i].distance(Px[j]);
                        min = new Tuple<>(Px[i], Px[j]);
                    }
            }
            return min;
        }

        int centerPoint = Px.length / 2;
        dPoint[] Qx = Arrays.copyOfRange(Px, 0, centerPoint);
        dPoint[] Rx = Arrays.copyOfRange(Px, centerPoint, Px.length);
        dPoint[] Qy = Arrays.copyOfRange(Py, 0, centerPoint);
        dPoint[] Ry = Arrays.copyOfRange(Py, centerPoint, Py.length);

        var minQ = getClosestPairRec(Qx, Qy);
        var minR = getClosestPairRec(Rx, Ry);

        var delta = Math.min(minQ.a.distance(minQ.b), minR.a.distance(minR.b));
        var L = Qx[Qx.length - 1];

        //dPoint[] S = Arrays.stream(gP).takeWhile(x -> L.distance(x) <= delta).toArray(dPoint[]::new);
        ArrayList<dPoint> S = new ArrayList<>();
        for (int i = 0; i < Px.length; i++) if (L.distance(Px[i]) <= delta) S.add(Px[i]);
        for (int i = 0; i < Py.length; i++) if (L.distance(Py[i]) <= delta) S.add(Py[i]);

        Tuple<dPoint> minS = null;
        double minvalue = Double.MAX_VALUE;
        for (int i = 0; i < S.size(); i++) {
            for (int j = i + 1; j <= i + 15 && j < S.size(); j++)
                if (S.get(i).distance(S.get(j)) < minvalue){
                    minvalue = S.get(i).distance(S.get(j));
                    minS = new Tuple<>(S.get(i), S.get(j));
                }
        }

        if (minS != null && minS.a.distance(minS.b) < delta)
            return minS;
        else if (minQ.a.distance(minQ.b) < minR.a.distance(minR.b))
            return minQ;
        else return minR;
    }
}


class dPoint {
    double x, y;

    public dPoint(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double distance(dPoint b) {
        return Math.sqrt(Math.pow(b.x - this.x, 2) + Math.pow(b.y - this.y, 2));
    }

    @Override
    public String toString() {
        return x + " " + y;
    }
}

class Tuple<T> {
    public final T a, b;

    public Tuple(T a, T b){
        this.a = a;
        this.b = b;
    }

    @Override
    public String toString() {
        return a.toString() + " " + b.toString();
    }
}
